adr/2025-04-01-context-gateway.md

```diff
@@ -0,0 +1,395 @@
+---
+title: Context Gateway  
+date: 2025-04-01
+area: checkout
+tags: [checkout, app, context, personalization, gateway, storefront]
+---
+
+# ADR: Context Gateway Feature
+
+![](https://md.sw-bench.de/uploads/645666dd-c45d-4198-b4f6-e04e1836dfdb.png)
+
+## Context
+
+Previously, we introduced the `CheckoutGateway` to allow app servers to influence the checkout process based on the current cart and sales channel context.  
+This was a significant step toward enabling dynamic, app-driven decision-making during checkout.
+
+However, this approach does not support modifying the storefront experience **outside** of the checkout flow.  
+There is a growing need to allow app servers to influence the **sales channel context** —
+e.g. customer data, language, and currency — based on external logic or user-specific criteria.
+
+In particular, some use cases require **context manipulation before a cart or checkout process begins**.  
+To address this, we propose the introduction of a dedicated `ContextGateway`,
+which provides a secure and structured communication channel between the storefront and the app server.
+
+The gateway is initiated from the storefront client and allows apps to request context changes based on external decision logic.
+
+### Example Use Cases
+
+- Registering customers with custom or prefilled data (including guest and full customer registration)
+- Logging in existing customers
+- Switching the current language or currency
+- Updating the active customer address
+
+## Decision
+
+### AppContextGateway
+
+To encapsulate the logic for context-driven decisions, a new `AppContextGateway` will be introduced.
+
+It receives a `ContextGatewayPayloadStruct`, which includes:
+
+- The current `SalesChannelContext`
+- The current `Cart`
+- A `RequestDataBag` containing all custom data provided by the client
+
+The gateway will return a `ContextTokenResponse`, which includes the updated token of the `SalesChannelContext` after executing all applicable commands.
+
+Example implementation:
+
+```php
+<?php declare(strict_types=1);
+
+class AppCheckoutGateway
+{
+    public function process(ContextGatewayPayloadStruct $payload): ContextTokenResponse;
+}
+```
+
+#### Store-API
+
+A new store API route, `ContextGatewayRoute` '/store-api/context/gateway', will be introduced.
+This route will call the `AppContextGateway` implementation and respond accordingly.
+
+Example implementation:
+
+```php
+<?php declare(strict_types=1);
+
+#[Route(defaults: ['_routeScope' => ['store-api']])]
+class ContextGatewayRoute extends AbstractContextGatewayRoute
+{
+    public function __construct(
+        private readonly AppContextGateway $contextGateway,
+    ) {
+    }
+
+    public function getDecorated(): AbstractContextGatewayRoute
+    {
+        throw new DecorationPatternException(self::class);
+    }
+
+    #[Route(path: '/store-api/context/gateway', name: 'store-api.context.gateway', methods: ['GET', 'POST'])]
```

Comment:
To invoke that endpoint App also needs to know sw-access-key - is it something the App know?

I'd consider put it under app-gateway, as it's super tailored to App system and might be confusing for someone who's browsing the API.

Reply:
@patzick 
The app would trigger this client-side via JS to the StorefrontController, which in turn calls the route.
In headless environments the app has to call admin api first and retrieve the SalesChannel's accessKey and call the route immediately.

Reply:
Yeah, but we did the same with the checkout gateway, so I try to keep it aligned for now:

see: https://github.com/shopware/shopware/blob/trunk/src/Core/Checkout/Gateway/SalesChannel/CheckoutGatewayRoute.php#L42

Reply:
Okay, now I understand from where is that coming from, can you point me for some documentation on how that one goes? 
Because basically your proposal would make `checkout gateway` unnecessary, as an app has user session context and can act on checkout routes in their name as well